### ISSUE TITLE
Making generic index type array optional breaks type safety

### ISSUE THREAD

**Issue by mfsjr** [Microsoft/TypeScript#20771]
I'm trying to create a generic function for updating an interface's arrays at an index.  
```ts
interface Person {
  nicknames: Array<string>
}

export interface Student extends Person {
  scores: Array<number>,
  schoolName: string
}

export let updateStudent = function<S extends Student, K extends keyof S, V extends keyof S[K]>
  (_student: S, propName: K, value: S[K][V], index: number) {
  _student[propName][index] = value;
};

let student: Student = {
  scores: [80, 90],
  nicknames: ['stinky', 'butthead'],
  schoolName: 'Central'
};

updateStudent(student, "scores", 'a', 0); // ERROR: TS2345
```
This error is what I would expect, I'm passing a string where a number is expected.

However, there is no error if you change the definition of 'scores' in the Student interface to:
`scores?: Array<number>`

**Expected behavior:**
Making 'scores' optional should not allow passing a string value to an array of numbers (or undefined array), ie, should get the error.

**Actual behavior:**
Making 'scores' optional allows you to pass the incorrect type without an error.

**Comment by RyanCavanaugh** [comment]
Context: Needs `strictNullChecks` on

**Comment by RyanCavanaugh** [comment]
@ahejlsberg thoughts? Simplified version:
```ts
export interface Student {
    scores: Array<number>; // When optional, no error occurs
    schoolName: string;
}

export let updateStudent = function<S extends Student, K extends keyof S, V extends keyof S[K]>
  (_student: S, propName: K, value: S[K][V]) {
  _student[propName][0] = value;
};

let student: Student = {
  scores: [80, 90],
  schoolName: 'Central'
};

updateStudent(student, "scores", 'a'); 
```

**Comment by mfsjr** [comment]
This problem appears to be fixed now in Playground, which I presume is 2.7?

**Comment by mhegazy** [comment]
thanks. closing.
